javalib/src/main/scala/java/lang/Number.scala

```diff
@@ -11,7 +11,9 @@ abstract class Number extends java.lang._Object with java.io.Serializable {
   def doubleValue(): scala.Double
 
   @inline override def __scala_==(other: _Object): scala.Boolean = {
-    if (other.isInstanceOf[ScalaNumber] && !this.isInstanceOf[ScalaNumber]) {
+    if (this.isInstanceOf[ScalaNumber]) {
+      this.equals(other)
+    } else if (other.isInstanceOf[ScalaNumber]) {
       other.equals(this)
     } else {
       super.__scala_==(other)
```

Comment:
I don't think this is the appropriate fix. The previous code *should* have been correct, if `java.lang.Object.__scala_==` had a reasonable implementation. The reasonable implementation should be:
```scala
  @inline def __scala_==(that: _Object): scala.Boolean =
    this.equals(that)
```
Instead, it's
https://github.com/scala-native/scala-native/blob/b49e74efe7079bdf415b5724da4acb87c5567210/nativelib/src/main/scala/java/lang/Object.scala#L39-L44
as an optimization based on some vtable hackery.

Unfortunately the vtable hackery doesn't patch up `super.__scala_==` calls. That means we can never do `super.__scala_==`, and instead we should inline the correct, reasonable implementation that is written nowhere.

So the correct implementation of `Number.__scala_==` should be:
```scala
  @inline override def __scala_==(other: _Object): scala.Boolean = {
    if (other.isInstanceOf[ScalaNumber] && !this.isInstanceOf[ScalaNumber]) {
      other.equals(this)
    } else {
      this.equals(other)
    }
  }
```

Reply:
I committed with the recommended fix and Travis CI is all green.  Thank you for the patient 
explanation.

Reply:
After the successful commit, I sat down and drew out the truth tables, mentally thanking my Discrete Math professor.

I believe the following code is equivalent and correct.  It passes the same tests as the
code in the current commit.   I find this code to be easier to read and understand.
I think future developers may also find it easier.

```
  @inline override def __scala_==(other: _Object): scala.Boolean = {           
    if ((this.isInstanceOf[ScalaNumber]) ||                                    
        (!other.isInstanceOf[ScalaNumber])) {                                  
      this.equals(other)                                                       
    } else {                                                                   
      other.equals(this)                                                       
    }       
```
It may also have performance benefits. I believe that the `this` left comparand will more
 likely be a ScalaNumber, by a large factor. The code above would do one test in that case,
leading to a lower amortized cost for some/most workloads.

Since the code is inlined, and the type of `this` is known at compile time, there is
a sporting chance that the inliner will reduce the call to `this.equals(other)` when
this is known to be a ScalaNumber.

Normally, I would go with the commited code and be done with it. However, this is a 
`hot` path, where likely instruction path length matters, as long as correctness is maintained.

Insert the classical refrain of "Show me the data, across a series of workloads'.

Note to future self:   

The test for `other` being a scala number is necessary to pass the 'j.l.Integer == BigInt` 
case and similar in the test suite.  The code above can not be correctly reduced to `this.equals(other)`.

Reply:
> I believe the following code is equivalent and correct. It passes the same tests as the
> code in the current commit. I find this code to be easier to read and understand.
> I think future developers may also find it easier.
> 
> I can do another commit if my proposed code is acceptable.
> 
> ```
>   @inline override def __scala_==(other: _Object): scala.Boolean = {           
>     if ((this.isInstanceOf[ScalaNumber]) ||                                    
>         (!other.isInstanceOf[ScalaNumber])) {                                  
>       this.equals(other)                                                       
>     } else {                                                                   
>       other.equals(this)                                                       
>     }       
> ```

This is indeed correct. I'm not sure about the performance impact.

> It may also have performance benefits. I believe that the `this` left comparand will more
> likely be a ScalaNumber, by a large factor. The code above would do one test in that case,
> leading to a lower amortized cost for some/most workloads.

The cases where `this` is a `ScalaNumber` are basically *only* for `scala.math.BigInt` and `scala.math.BigDecimal`. For those, we know that equality is expensive (at least it's more than one or two tests) so their equality will dwarf whatever optimization we try to do for them in `__scala_==`. Therefore, this does not seem to be a scenario worth optimizing for.

> Since the code is inlined, and the type of `this` is known at compile time, there is
> a sporting chance that the inliner will reduce the call to `this.equals(other)` when
> this is known to be a ScalaNumber.

That is true, but the same reasoning applies to the existing code as well. If `this` is known to be a `ScalaNumber` at compile time, then the condition
```scala
other.isInstanceOf[ScalaNumber] && !this.isInstanceOf[ScalaNumber]
```
reduces to
```scala
other.isInstanceOf[ScalaNumber] && false
```
which, because `isInstanceOf` is always pure, can also be optimized down to
```scala
false
```
allowing the whole method to reduce to `this.equals(that)`.